loader/loader.go

```diff
@@ -1,5 +1,165 @@
 package loader
 
-// Loader represents loading process from a bitcoin client.
+import (
+	"fmt"
+
+	"github.com/IlliniBlockchain/etl-bitcoin/client"
+	"github.com/IlliniBlockchain/etl-bitcoin/database"
+	"github.com/IlliniBlockchain/etl-bitcoin/types"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+)
+
+// General idea for loader:
+// Represents a single entire process of getting data from a source to a destination.
+// It takes in clients (RPC or DB) and implements the loading process for transporting
+// data between them.
+
+// LoaderOptions represents a map of arbitrary options when constructing a loader.
+type LoaderOptions map[string]interface{}
+
+// ClientDatabaseLoader represents any process of loading data from a
+// Client to a Database.
 type Loader interface {
+	Run(BlockRange, database.DBTx)
+	Close()
+}
+
+type ClientDBLoaderConstructor func(client client.Client, db database.Database, opts LoaderOptions) (Loader, error)
+
+type ClientDBLoader struct {
+	client     client.Client
+	db         database.Database
+	maxWorkers int
+	msgs       chan LoaderMsg
+}
+
+type LoaderMsg struct {
+	t          LoaderMsgType
+	blockRange BlockRange
+	dbTx       *database.DBTx
+	data       interface{}
+}
+
+// Message type enum
+type LoaderMsgType int
+
+const (
+	Range LoaderMsgType = iota
+	Hashes
+	Blocks
+	BlocksForHeaders
+	BlocksForTxs
+	Headers
+	Txs
+)
+
+type BlockRange struct {
+	startBlockHeight int64
+	endBlockHeight   int64
+}
+
+func NewClientDBLoader(client client.Client, db database.Database, opts LoaderOptions, maxWorkers int) (*ClientDBLoader, error) {
+	// initialize struct
+	msgs := make(chan LoaderMsg)
+	loader := &ClientDBLoader{
+		client:     client,
+		db:         db,
+		msgs:       msgs,
+		maxWorkers: maxWorkers,
+	}
+
+	// start workers
+	for i := 0; i < maxWorkers; i++ {
+		go loader.loaderWorker()
```

Comment:
Consider using a [wait group](https://pkg.go.dev/sync#WaitGroup) here so that when we are done with the loader and want it to stop, we have a way to gracefully wait for it to finish. Maybe even better here, is to use an [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup) like I did [here](https://github.com/IlliniBlockchain/etl-bitcoin/pull/16/files#diff-66034a274276421c4c95008dc3eea5bc9fb472af7c76403afce8c9be501c1bb8R35-R49). An error group is nice because it'll do all the error handling for you.

Reply:
Ah cool. Was wondering the difference between wait and error group. Makes sense.